### PR TITLE
Fix 330 build error and add 322 shims layer [databricks]

### LINKED ISSUE #4438: [FEA] Create shims for spark 3.2.2
**Is your feature request related to a problem? Please describe.**
https://github.com/apache/spark/commit/fd7bd3ef6b7635ce35b4cc35362d814d5204cd27#diff-9c5fb3d1b7e3b0f54bc5c4182965c4fe1f9023d449017cece3005d3f90e8e4d8

Spark 3.2.1 snapshot stoped updating and have no release 3.2.1, currently it changed to 3.2.2 snapshot.
Let's keep 3.2.1 snapshot shims and create 3.2.2 snapshot shims.
Later if confirm 3.2.1 snapshot is useless, then we can remove 3.2.1 snapshot.

### PR DESCRIPTION
**Pull request by res-life**
This fixes #4438, #4432 and #4443

Co-authored-by: Firestarman <firestarmanllc@gmail.com>
Co-authored-by: Kuhu Shukla <kuhus@nvidia.com>

Compile is broken for builver=330 after https://issues.apache.org/jira/browse/SPARK-37705, And it also impacted the build for Spark 3.2.2 snapshot. So fix the 2 issues togother in one PR.

### CODE REVIEW

**Comment by res-life** [comment]
Spark 3.2.1 snapshot stoped updating and have no release 3.2.1, currently it changed to 3.2.2 snapshot.
Let's keep 3.2.1 snapshot shims and create 3.2.2 snapshot shims.
Later if confirm 3.2.1 snapshot is useless, then we can remove 3.2.1 snapshot.

**Comment by firestarman** [comment]
build

**Comment by res-life** [comment]
This [PR](https://github.com/NVIDIA/spark-rapids/pull/4444) is also about it. @kuhushukla Can you help to review?

**Comment by firestarman** [comment]
build

**Comment by kuhushukla** [comment]
The changes prima facie look fine. We could probably name the 322+ and 320+ appropriately and more precisely in order to not confuse what they really mean. At this point we have three such folders including 330+.

**Comment by kuhushukla** [comment]
> This [PR](https://github.com/NVIDIA/spark-rapids/pull/4444) is also about it. @kuhushukla Can you help to review?

It would be nice if you could combine the two PRS and add a coauthor.

**Comment by res-life** [comment]
build

File: sql-plugin/src/main/330+/scala/com/nvidia/spark/rapids/shims/v2/GpuHashPartitioning.scala [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+    super.satisfies0(required) || {
+      required match {
+        case ClusteredDistribution(requiredClustering, _) =>
+          expressions.forall(x => requiredClustering.exists(_.semanticEquals(x)))
+        case _ => false
+      }
+    }
+  }
+
+ // override def createShuffleSpec(distribution: ClusteredDistribution): ShuffleSpec =
```
Comment (gerashegalov) [inline sql-plugin/src/main/330+/scala/com/nvidia/spark/rapids/shims/v2/GpuHashPartitioning.scala:37]: remove code commented out

File: sql-plugin/src/main/330+/scala/com/nvidia/spark/rapids/shims/v2/GpuShuffledHashJoinExec.scala [hunk 2]
```diff
@@ -90,9 +91,10 @@ case class GpuShuffledHashJoinExec(
...
     condition,
-    isSkewJoin = isSkewJoin,
-    cpuLeftKeys,
-    cpuRightKeys) {
+    isSkewJoin = isSkewJoin) {
 
   override def otherCopyArgs: Seq[AnyRef] = cpuLeftKeys :: cpuRightKeys :: Nil
+
+  override def requiredChildDistribution: Seq[Distribution] =
+    ClusteredDistribution(cpuLeftKeys) :: ClusteredDistribution(cpuRightKeys) :: Nil
```
Comment (gerashegalov) [inline sql-plugin/src/main/330+/scala/com/nvidia/spark/rapids/shims/v2/GpuShuffledHashJoinExec.scala:99]: nit: `Seq(ClusteredDistribution(cpuLeftKeys), ClusteredDistribution(cpuRightKeys))`

**Review by gerashegalov — APPROVED** [review]
This looks good to me. I have mostly nits and some ideas to minimizing the patch but it's ok to address it in a follow-on.

**Review by kuhushukla — DISMISSED** [review]

**Comment by sameerz** [comment]
build

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/SparkShims.scala [hunk 3]
```diff
@@ -109,7 +108,8 @@ trait SparkShims {
     pushDownStartWith: Boolean,
     pushDownInFilterThreshold: Int,
     caseSensitive: Boolean,
-    datetimeRebaseMode: LegacyBehaviorPolicy.Value): ParquetFilters
+    lookupFileMeta: String => String,
+    modeByConfig: String): ParquetFilters
```
[the remark this answers is not in the document]
Comment (revans2) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/SparkShims.scala:112]: Yes totally agree these names are too generic. That or we need to add scala docs to explain what they do.

File: sql-plugin/src/main/320until322/scala/com/nvidia/spark/rapids/shims/v2/Spark320until322Shims.scala [hunk 4]
```diff
@@ -0,0 +1,1123 @@
...
+  override def getBuildSide(join: HashJoin): GpuBuildSide = {
+    GpuJoinUtils.getGpuBuildSide(join.buildSide)
+  }
+
+  override def getBuildSide(join: BroadcastNestedLoopJoinExec): GpuBuildSide = {
+    GpuJoinUtils.getGpuBuildSide(join.buildSide)
+  }
+
+  override def getDateFormatter(): DateFormatter = {
+    // TODO verify
```
Comment (revans2) [inline sql-plugin/src/main/320until322/scala/com/nvidia/spark/rapids/shims/v2/Spark320until322Shims.scala:179]: Is this TODO done? I see the same thing in the original Spark32XShims.scala so I am fine if this is a follow on issue. We really don't want a TODO like this in our code.

**Review by revans2 — APPROVED** [review]
Overall I just have nits. I would like to at least see a follow on issue to address the names of the new parameters for the parquet filter shim layer and something to look into the TODO comment. Even if it ends up being that we just delete the comment.

**Comment by res-life** [comment]
build

**Comment by firestarman** [comment]
LGTM

**Review by GaryShen2008 — APPROVED** [review]
Approved for the last commit for some nit changes.
The previous commits have been approved. Let's merge it.

**Comment by kuhushukla** [comment]
@res-life can you add the follow on issues here and assign them to yourself if you are working on them. We can also assign it to someone else if you would prefer.

**Pull request merged**